### PR TITLE
Add practitioner query param functionality to $evaluate-measure

### PR DESCRIPTION
**Pull request by sarahmcdougall**
# Summary
This PR implements the practitioner query param for the `$evaluate-measure` operation.

## New behavior
The test server now accepts a `practitioner` ID as a query parameter when invoking the `$evaluate-measure` operation.

According to the `$evaluate-measure` [spec](https://www.hl7.org/fhir/operation-measure-evaluate-measure.html), the practitioner query parameter has cardinality `0..1`. In contrast to the [$care-gaps operation](https://build.fhir.org/ig/HL7/davinci-deqm/OperationDefinition-care-gaps.html), the subject and practitioner query parameters are **not** mutually exclusive for $evaluate-measure. If the practitioner is specified for the measure that will be calculated, the measure will be calculated **only for subjects that have a primary relationship to the identified practitioner**.

## Code changes
Code changes are inspired by [this PR that adds support for the practitioner parameter for the $care-gaps operation](https://github.com/projecttacoma/deqm-test-server/pull/93).

Most code changes appear in `measure.service.js` within `evaluateMeasure()`. Now, the function checks for the presence of the practitioner query param. If present, the function narrows down the patients to calculate the measure against so that the calculation only occurs for patients that reference the practitioner.

Unit tests were added to reflect the new query parameter.

# Testing guidance
Run `npm test` to ensure that the new unit tests pass and that all previous tests still pass. Check that the unit tests cover all possible situations that need to check for a practitioner (ex. A single Patient subject/individual report type, Group subject/population report type, etc.)

Steps for testing the new functionality in the test server:
* Run `npm run db:reset`, followed by `npm run connectathon-upload`
* Send a PUT request to `[http://localhost:3000/4_0_1/Practitioner/1`](http://localhost:3000/4_0_1/Practitioner/1%60) with the following body: 
```
{
	“resourceType”: “Practitioner”,
	“id”: “1”
}
```
* Modify some patients to reference this test practitioner. For example, we can first modify the denominator EXM130 patient to reference this practitioner. Send a PUT request to `[http://localhost:3000/4_0_1/Patient/denom-EXM130`](http://localhost:3000/4_0_1/Patient/denom-EXM130%60) with the following body:
```
{
	"resourceType": "Patient",
	"id": "denom-EXM130",
	“generalPractitioner”: {
		"identifier": {
			"value": “Practitioner/1”
		}
	}
}
```
* Send a GET request to `http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=denom-EXM130&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1`. You should receive the typical output for $evaluate-measure with the denominator patient, since the denominator patient references the practitioner
* Now, send a GET request to `http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=numer-EXM130&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1`.  Since this subject does not reference the practitioner, the measure will not be calculated for this subject. The output should reflect that.
* Try sending a request that omits the subject query parameter. In this case, in alignment with the spec, we should only have the measure calculated on the denominator patient again since it is the only patient that references our practitioner.
* Try sending a request with a Group. See [this PR](https://github.com/projecttacoma/deqm-test-server/pull/86) for an example Group resource with EXM130 patients. If using this Group, you can send a request to http://localhost:3000/4_0_1/Measure/measure-EXM130-7.3.000/$evaluate-measure?subject=Group/EXM130-patients&periodStart=2019-01-01&periodEnd=2019-12-31&practitioner=Practitioner/1&reportType=population. Since the Group contains the numerator and denominator patients but only the denominator patient references the practitioner, the measure report should reflect this.
* Try sending other requests (have multiple patients reference the practitioner, etc.)

### CODE REVIEW

File: src/util/patientUtils.js [hunk 1]
```diff
@@ -106,4 +108,30 @@ const retrievePatientIds = async ({ subject, organization, practitioner }) => {
...
+ */
+const filterPatientIdsFromGroup = async (group, practitioner) => {
+  const patientPromises = group.member.map(async m => {
+    const query = {
+      id: m.entity.reference.split('/')[1],
+      'generalPractitioner.identifier.value': practitioner
+    };
+    return await findOneResourceWithQuery(query, 'Patient');
+  });
+  const patients = (await Promise.all(patientPromises)).filter(a => a);
```
Comment (natjoe4) [inline src/util/patientUtils.js:127]: I think this line could use a comment. Took me a bit to figure out that this was filtering out the potentially empty search results. Or, consider changing this map + filter to a reduce

File: src/util/validationUtils.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ const { BadRequestError, NotImplementedError } = require('./errorUtils');
  */
 function validateEvalMeasureParams(query) {
   const REQUIRED_PARAMS = ['periodStart', 'periodEnd'];
-  const UNSUPPORTED_PARAMS = ['measure', 'practitioner', 'lastReceivedOn'];
+  const UNSUPPORTED_PARAMS = ['measure', 'lastReceivedOn'];
 
   checkRequiredParams(query, REQUIRED_PARAMS, '$evaluate-measure');
```
Comment (natjoe4) [inline src/util/validationUtils.js:13]: We should add a check here like we have in the gaps validation to make sure that Practitioner is is of the form Practitioner/id, and a test that the validation utils throw the right error in that case

**Review by natjoe4 — CHANGES_REQUESTED** [review]
All small changes to fix a few functionality bugs. Nice work!

File: src/services/measure.service.js [hunk 3]
```diff
@@ -225,49 +229,114 @@ const evaluateMeasure = async (args, { req }) => {
...
 
   if (req.query.reportType === 'population') {
-    let patientBundles = [];
-    if (req.query.subject) {
-      const subjectReference = req.query.subject.split('/');
-      const group = await findResourceById(subjectReference[1], subjectReference[0]);
-      if (!group) {
-        throw new ResourceNotFoundError(
-          `No resource found in collection: ${subjectReference[0]}, with: id ${subjectReference[1]}.`
+    await evaluateMeasureForPopulation(args, { req });
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply src/services/measure.service.js]: did ya one better ;)

File: src/services/measure.service.js [hunk 4]
```diff
@@ -225,49 +229,114 @@ const evaluateMeasure = async (args, { req }) => {
...
+  } else {
+    await evaluateMeasureForIndividual(args, { req });
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply src/services/measure.service.js]: did ya one better ;)

File: src/util/patientUtils.js [hunk 5]
```diff
@@ -106,4 +108,30 @@ const retrievePatientIds = async ({ subject, organization, practitioner }) => {
...
+ * as their generalPractitioner will be selected for gaps calculation run on them
+ * @returns array of patient IDs
+ */
+const filterPatientIdsFromGroup = async (group, practitioner) => {
+  const patientPromises = group.member.map(async m => {
+    const query = {
+      id: m.entity.reference.split('/')[1],
+      'generalPractitioner.identifier.value': practitioner
+    };
+    return await findOneResourceWithQuery(query, 'Patient');
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply src/util/patientUtils.js]: Applied in [b00efd4](https://github.com/projecttacoma/deqm-test-server/pull/96/commits/b00efd44ab832098a8c2fbbc974ad9b0b90f16eb) with other changes

File: src/util/patientUtils.js (hunk 1 above)
Comment (mgramigna) [reply src/util/patientUtils.js]: [3cc81eb](https://github.com/projecttacoma/deqm-test-server/pull/96/commits/3cc81eb3436cb880f935f0bb48a6cbd6eee513c9)

File: src/util/validationUtils.js (hunk 2 above)
Comment (mgramigna) [reply src/util/validationUtils.js:13]: [3cc81eb](https://github.com/projecttacoma/deqm-test-server/pull/96/commits/3cc81eb3436cb880f935f0bb48a6cbd6eee513c9)

**Review by natjoe4 — APPROVED** [review]
Looks good to me!

**Pull request merged**